src/main/scala/mesosphere/marathon/raml/EnvVarConversion.scala

```diff
@@ -8,17 +8,17 @@ import scala.collection.immutable.Map
 trait EnvVarConversion {
   implicit val envVarRamlWrites: Writes[Map[String, state.EnvVarValue], Map[String, EnvVarValueOrSecret]] =
     Writes {
-      _.map {
-        case (name, state.EnvVarString(v)) => name -> EnvVarValue(v)
-        case (name, state.EnvVarSecretRef(secret: String)) => name -> EnvVarSecret(secret)
+      _.mapValues {
```

Comment:
Nice catch! What do you think about  using the builder from [Map.newBuilder](http://www.scala-lang.org/api/2.12.3/scala/collection/immutable/Map$.html#newBuilder[A,B]:scala.collection.mutable.Builder[(A,B),CC[A,B]]) or `breakOut` instead?

```scala
implicit val envVarRamlWrites: Writes[Map[String, state.EnvVarValue], Map[String, EnvVarValueOrSecret]] = new Writes {
  override def write(in: Map[String, state.EnvVarValue]): Map[String, EnvVarValueOrSecret] = {
    val builder = Map.newBuilder[String, EnvVarValueOrSecret]
    in.foreach {
      case (name, state.EnvVarString(v))=> builder += key -> EnvVarValue(v)
      case (name, state.EnvVarSecretRef(secret: String)) => builder += name -> EnvVarSecret(secret)
    }
    builder.result()
  }
}
```

or

```scala
implicit val envVarRamlWrites: Writes[Map[String, state.EnvVarValue], Map[String, EnvVarValueOrSecret]] = new Writes {
  override def write(in: Map[String, state.EnvVarValue]): Map[String, EnvVarValueOrSecret] = {
    in.iterator.map {
      case (name, state.EnvVarString(v))=> key -> EnvVarValue(v)
      case (name, state.EnvVarSecretRef(secret: String)) => name -> EnvVarSecret(secret)
    }.breakOut
  }
}
```

Reply:
Either `Map.newBuilder` or `breakOut` would construct a new Map from scratch. That'd involve rehashing all the keys and the additional memory allocation we're aiming to avoid.

The important bits of .mapValues are here: https://github.com/scala/scala/blob/v2.12.4/src/library/scala/collection/MapLike.scala#L252-L265. Only downside would be if the function is expensive or invoked for the same key many times--neither happens here.

Reply:
Good point. I'd like to see some benchmarks. We did see issue with `mapValues` being lazy in the past. It is nicer for the GC to make copies.

Reply:
I've done a quick test with `breakOut` and a builder. Both are slightly slower than this solution. So it's good with me. We probably could achieve more speed in the serialization as well. Further more a lighter response should help as well.

Reply:
> We probably could achieve more speed in the serialization as well.

I have some ideas around this, but they're not the low hanging fruit that this PR is.

If we could skip the AST step of case class => AST => String (e.g. scala-pickling or sbt-serialization), that'd cut out a good chunk of computation, but it'd involve rewriting that 1,000-line RAML code generator. I'm afraid to touch that.

For remaining optimizations of the current approach, play.json...JsObject uses a mutable.LinkedHashMap internally. I think we could get decent performance gains with a more optimized Map implementation (e.g. guava), but that'd involve hooking into the JsObject package private constructor. If we wanted to avoid touching the Raml generator, we could hook in the replacement with AspectJ since you're already using it with kamon.

If you think performance gains could justify some dark magic, I could through up another PR and benchmarks. Otherwise, if using package private constructors is a deal-breaker, I'll leave it be.

Let me know how you feel, or if you have other optimizations in mind. We're feeling the performance pain, so I wouldn't mind donating some time to it, especially with how generous you've all been with PR reviews and merges!

Reply:
Correction: The JsObject [*field* is private, not the constructor](https://github.com/playframework/play-json/blob/2.6.7/play-json/shared/src/main/scala/play/api/libs/json/JsValue.scala#L125). This may be less dark and dirty than I previously thought.